### PR TITLE
Am/feat/semver trick

### PR DESCRIPTION
**Pull request by IceTDrinker**
0.4.2 forward compat code

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -263,17 +270,25 @@ build_tfhe_full: install_rs_build_toolchain
 	RUSTFLAGS="$(RUSTFLAGS)" cargo $(CARGO_RS_BUILD_TOOLCHAIN) build --profile $(CARGO_PROFILE) \
 		--features=$(TARGET_ARCH_FEATURE),boolean,shortint,integer -p $(TFHE_SPEC) --all-targets
 
+.PHONY: symlink_c_libs_without_fingerprint # Link the .a and .so files without the changing hash part in target
+symlink_c_libs_without_fingerprint:
+	@./scripts/symlink_c_libs_without_fingerprint.sh \
```
Comment (soonum) [inline Makefile:275]: We don't want the output of the script ?

File: Makefile [hunk 2]
```diff
@@ -333,7 +348,8 @@ test_c_api_rs: install_rs_check_toolchain
 		c_api
 
 .PHONY: test_c_api_c # Run the C tests for the C API
-test_c_api_c: build_c_api
+test_c_api_c:
+	@FORWARD_COMPAT=forward_compatibility "$(MAKE)" build_c_api
```
Comment (soonum) [inline Makefile:352]: Is there no case in which we want to run tests on C API without the forward compatibility?

File: tfhe/Cargo.toml [hunk 3]
```diff
@@ -78,29 +82,49 @@ getrandom = { version = "0.2.8", optional = true }
...
-    "wasm-bindgen",
-    "js-sys",
-    "console_error_panic_hook",
-    "serde-wasm-bindgen",
-    "getrandom",
+    "dep:wasm-bindgen",
+    "dep:js-sys",
+    "dep:console_error_panic_hook",
+    "dep:serde-wasm-bindgen",
+    "dep:getrandom",
```
Comment (soonum) [inline tfhe/Cargo.toml:127]: I don't understand why there are dependencies with `dep:` prefix and others not.
Can I have the rationale behind or link toward this feature ?

File: tfhe/docs/core_crypto/tutorial.md [hunk 4]
```diff
@@ -9,7 +9,7 @@ Welcome to this tutorial about `TFHE-rs` `core_crypto` module.
 To use `TFHE-rs`, it first has to be added as a dependency in the `Cargo.toml`:
 
 ```toml
-tfhe = { version = "0.4.1", features = [ "x86_64-unix" ] }
+tfhe = { version = "0.4.2", features = [ "x86_64-unix" ] }
```
Comment (soonum) [inline tfhe/docs/core_crypto/tutorial.md:12]: I assume it's a backport. This would be changed for `"0.5.0"` for the new release ?

**Review by soonum — COMMENTED** [review]
Yay :tada: the forward compatibility is almost there.
I just have some minor comments regarding the CI.

File: scripts/integer-tests.sh [hunk 5]
```diff
@@ -9,13 +9,15 @@ function usage() {
     echo "--rust-toolchain          The toolchain to run the tests with default: stable"
     echo "--multi-bit               Run multi-bit tests only: default off"
     echo "--cargo-profile           The cargo profile used to build tests"
+    echo "--tfhe-package            The package spec like tfhe@0.4.2, default=tfhe"
```
Comment (mayeul-zama) [inline scripts/integer-tests.sh:12]: Unrelated but could be fixed here, line 6 says shortint instead of integer

File: scripts/symlink_c_libs_without_fingerprint.sh [hunk 6]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env bash
+
+set -e
+
+function usage() {
+    echo "$0: symlink C libs to names without the variable fingerprint part"
```
Comment (mayeul-zama) [inline scripts/symlink_c_libs_without_fingerprint.sh:6]: What is the variable fingerprint part?

File: tfhe/docs/core_crypto/tutorial.md (hunk 4 above)
Comment (soonum) [reply tfhe/docs/core_crypto/tutorial.md:12]: My bad, I didn't read the content of the PR :zany_face:

File: tfhe/Cargo.toml (hunk 3 above)
Comment (mayeul-zama) [reply tfhe/Cargo.toml:127]: https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

File: Makefile (hunk 1 above)
Comment (IceTDrinker) [reply Makefile:275]: we will get the output the @ just silences the commands we launch

File: Makefile (hunk 2 above)
Comment (IceTDrinker) [reply Makefile:352]: no you are right it would be better, I'll update

File: tfhe/Cargo.toml (hunk 3 above)
Comment (IceTDrinker) [reply tfhe/Cargo.toml:127]: so normally all deps are now prefixed with dep: and features are not prefixed with dep:

it's something new-ish from cargo to avoid collisions between dependencies and features I believe

File: scripts/integer-tests.sh (hunk 5 above)
Comment (IceTDrinker) [reply scripts/integer-tests.sh:12]: yep will do

File: scripts/symlink_c_libs_without_fingerprint.sh (hunk 6 above)
Comment (IceTDrinker) [reply scripts/symlink_c_libs_without_fingerprint.sh:6]: like in the target/deps dir we get libthfe_c_api_dynamic_buffer-somehash.a so if we don't symlink to a "constant" name our builds would be a nightmare to make work, so I look for the lib that has the same name and is the most recent and say it's the lib we want

File: tfhe/src/lib.rs [hunk 7]
```diff
@@ -69,3 +69,7 @@ pub mod safe_deserialization;
 pub mod conformance;
 
 pub mod named;
+
+#[cfg(feature = "forward_compatibility")]
+/// cbindgen:ignore
+pub mod forward_compatibility;
```
Comment (mayeul-zama) [inline tfhe/src/lib.rs:75]: Could be part of a separate commit

File: tfhe/c_api_tests/test_high_level_integers.c [hunk 8]
```diff
@@ -171,6 +171,87 @@ int uint8_serialization(const ClientKey *client_key) {
...
+  assert(ok == 0);
+
+  ok = fhe_uint8_try_encrypt_with_client_key_u8(lhs_clear, deserialized_client_key, &lhs);
+  assert(ok == 0);
+
+  ok = fhe_uint8_serialize(lhs, &value_buffer);
+  assert(ok == 0);
+
+  deser_view.pointer = value_buffer.pointer;
+  deser_view.length = value_buffer.length;
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/c_api_tests/test_high_level_integers.c:213]: backlog issue I would say

File: tfhe/src/lib.rs (hunk 7 above)
Comment (IceTDrinker) [reply tfhe/src/lib.rs:75]: not really wanting to hunt down to make commits that are self consistent

File: tfhe/src/c_api/high_level_api/integers.rs [hunk 9]
```diff
@@ -357,6 +356,257 @@ create_integer_wrapper_type!(name: FheUint64, clear_scalar_type: u64);
 create_integer_wrapper_type!(name: FheUint128, clear_scalar_type: U128);
 create_integer_wrapper_type!(name: FheUint256, clear_scalar_type: U256);
 
+#[cfg(feature = "forward_compatibility")]
+pub mod forward_compatibility {
+    use super::*;
+
+    // FheInt don't have the 10, 12, 14 variants so we define the impl here
+    impl_update_serialization_format_on_type!(FheUint8);
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/c_api/high_level_api/integers.rs:364]: Maybe they are, but as we drop support for them I don't think spending more time and more specialized code for them is worth it

File: tfhe/src/c_api/high_level_api/integers.rs [hunk 10]
```diff
@@ -357,6 +356,257 @@ create_integer_wrapper_type!(name: FheUint64, clear_scalar_type: u64);
 create_integer_wrapper_type!(name: FheUint128, clear_scalar_type: U128);
 create_integer_wrapper_type!(name: FheUint256, clear_scalar_type: U256);
 
+#[cfg(feature = "forward_compatibility")]
+pub mod forward_compatibility {
+    use super::*;
+
+    // FheInt don't have the 10, 12, 14 variants so we define the impl here
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/c_api/high_level_api/integers.rs:363]: yes it's the same issue

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/c_api/high_level_api/integers.rs:363]: as the macro is shared if one kind of integer (signed or unsigned) cannot use the macro then the manual work has to be done for both kinds

[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/c_api/high_level_api/integers.rs:363]: Ok

File: tfhe/c_api_tests/test_high_level_integers.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/c_api_tests/test_high_level_integers.c:213]: I mean defining it in this file 
`BufferView take_view(Buffer input) {
  BufferView output;
  output.pointer = input.pointer;
  output.length = input.length;
}`
And then use it to remove `deser_view` variable

It's a very simple change which IMO is worth doing now

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/c_api_tests/test_high_level_integers.c:213]: problem is it's used everywhere else in the C tests and the objective is to get the 0.4.2 out the door, I don't want to fight with C now, so back log issue

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/c_api_tests/test_high_level_integers.c]: https://github.com/zama-ai/tfhe-rs-internal/issues/423

**Comment by IceTDrinker** [comment]
ah wait, the if def is not handled properly for the forward compat stuff

**Comment by IceTDrinker** [comment]
alright, could confirm the test properly compiles and runs ok, let's wait for the pcc checks but should be good

File: Makefile (hunk 2 above)
Comment (IceTDrinker) [reply Makefile]: this little maneuver cost us 51 years 😂 but it was necessary

**Comment by IceTDrinker** [comment]
the red status is the cancelled M1 tests, on 0.4 we still had the release lto off profile for it, let's not wait 6 hours to see the mac fail

**Comment by IceTDrinker** [comment]
alright looks like pcc is passing, good to review ! :)

File: tfhe/c_api_tests/test_forward_compatibility.c [hunk 11]
```diff
@@ -0,0 +1,120 @@
+#include <tfhe.h>
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+
+#ifdef DEFINE_FORWARD_COMPATIBILITY
```
Comment (tmontaigu) [inline tfhe/c_api_tests/test_forward_compatibility.c:7]: very minor thing: I would prefer `WITH_FEATURE_FORWARD_COMPATIBILITY`, the `#ifdef DEFINE` feels like a repetition to me

**Review by soonum — COMMENTED** [review]
CI looks good to me.

File: tfhe/c_api_tests/test_forward_compatibility.c (hunk 11 above)
Comment (IceTDrinker) [reply tfhe/c_api_tests/test_forward_compatibility.c:7]: ok will rename then, it was just the cbindgen style

**Review by mayeul-zama — APPROVED** [review]
Thanks!

**Review by tmontaigu — APPROVED** [review]

**Comment by IceTDrinker** [comment]
@slab-ci cpu_wasm_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_wasm_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_wasm_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_wasm_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_wasm_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_wasm_test

**Comment by IceTDrinker** [comment]
integer run to wait https://github.com/zama-ai/tfhe-rs/actions/runs/7627088419
